atef/procedure.py

```diff
@@ -878,20 +924,335 @@ def from_origin(
+            return Result(
+                severity=Severity.error,
+                reason=('One or more success criteria failed to initialize: '
+                        f'{[check.name for check in self.prepared_checks_failures]}')
+            )
+
+        severity = _summarize_result_severity(GroupResultMode.all_,
+                                              check_results + plan_results)
+        reason = (f'{len(plan_results)} plans run, '
+                  f'{len(check_results)} checks passed')
+        return Result(severity=severity, reason=reason)
+
+    @classmethod
+    def from_origin(
+        cls,
+        origin: PlanStep,
+        parent: Optional[PreparedProcedureGroup] = None
+    ) -> PreparedPlanStep:
+
+        prep_step = cls(
+            origin=origin,
+            parent=parent,
+            name=origin.name,
+        )
+
+        for plan_step in origin.plans:
+            try:
+                prep_plan = PreparedPlan.from_origin(plan_step, parent=prep_step)
+                prep_step.prepared_plans.append(prep_plan)
+            except Exception:
+                prep_step.prepared_plan_failures.append(plan_step)
+
+        for check in origin.checks:
+            res = create_prepared_comparison(check, parent=prep_step)
+            if isinstance(res, Exception):
+                prep_step.prepared_checks_failures.append(res)
+            else:
+                prep_step.prepared_checks.append(res)
+
+        return prep_step
+
+
+def create_prepared_comparison(
+    check: Union[ComparisonToTarget, ComparisonToPlanData],
+    parent: Optional[AnyDataclass] = None
+) -> Union[PreparedComparison, PreparedComparisonException]:
+
+    output = ValueError('Cannot prepare the provided comparison, type not '
+                        f'supported: ({check})')
+
+    if isinstance(check, ComparisonToTarget):
+        signal = check.to_signal()
+        comp = check.comparison
+        try:
+            output = PreparedSignalComparison.from_signal(
+                signal=signal, comparison=comp, parent=parent
+            )
+        except Exception as ex:
+            output = PreparedComparisonException(
+                exception=ex,
+                identifier=getattr(signal, 'pvname', ''),
+                message='Failed to initialize comparison',
+                comparison=comp,
+                name=comp.name
+            )
+
+    elif isinstance(check, ComparisonToPlanData):
+        comp = check.comparison
+        try:
+            output = PreparedPlanComparison.from_comp_to_plan(
+                check, parent=parent
+            )
+        except Exception as ex:
+            output = PreparedComparisonException(
+                exception=ex,
+                identifier=getattr(check, 'plan_id', ''),
+                message='Failed to initialize comparison',
+                comparison=comp,
+                name=comp.name
+            )
+
+    return output
+
+
+def get_bs_state(dclass: Any):
```

Comment:
Docstring and annotation would be nice here (not entirely obvious from the name)